### PR TITLE
Add support for Lacrosse TX34 rain gauge

### PR DESCRIPTION
**Pull request by rey42**

### CODE REVIEW

File: src/devices/lacrosse_tx34.c [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    Copyright (C) 2021 Reynald Poittevin
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+*/
+
+/**
+   LaCrosse TX34-IT rain gauge.
```
Comment (zuckschwerdt) [inline src/devices/lacrosse_tx34.c:13]: Don't indent in this comment

File: src/devices/lacrosse_tx34.c [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+#define LACROSSE_TX34_CRC_POLY 0x31
+#define LACROSSE_TX34_CRC_INIT 0x00
+#define LACROSSE_TX34_PREAMBLE_BITS 22
+#define LACROSSE_TX34_ITMODEL 5
+#define LACROSSE_TX34_PAYLOAD_BITS 40
+#define LACROSSE_TX34_RAIN_FACTOR 0.222
+
+static int lacrosse_tx34_callback(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+  data_t *data;
```
Comment (zuckschwerdt) [inline src/devices/lacrosse_tx34.c:56]: Indent with 4 spaces please. See https://triq.org/rtl_433/CONTRIBUTING.html#code-style

File: src/devices/lacrosse_tx34.c [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+static char *output_fields[] =
+  {
+    "model",
+    "id",
+    "battery_ok",
+    "newbattery",
+    "rain_mm",
+    "rain_raw",
+    "mic",
+    NULL
```
Comment (zuckschwerdt) [inline src/devices/lacrosse_tx34.c:144]: add a trailing `,`

File: src/devices/lacrosse_tx34.c [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+r_device lacrosse_tx34 = {
+    .name           = "LaCrosse TX34-IT rain gauge",
+    .modulation     = FSK_PULSE_PCM,
+    .short_width    = 58,
+    .long_width     = 58,
+    .reset_limit    = 4000,
+    .decode_fn      = &lacrosse_tx34_callback,
+    .disabled       = 0,
+    .fields         = output_fields
```
Comment (zuckschwerdt) [inline src/devices/lacrosse_tx34.c:155]: add a trailing `,`

File: src/devices/lacrosse_tx34.c (hunk 1 above)
Comment (rey42) [reply src/devices/lacrosse_tx34.c]: done

File: src/devices/lacrosse_tx34.c (hunk 2 above)
Comment (rey42) [reply src/devices/lacrosse_tx34.c]: done

File: src/devices/lacrosse_tx34.c (hunk 3 above)
Comment (rey42) [reply src/devices/lacrosse_tx34.c]: done

File: src/devices/lacrosse_tx34.c (hunk 4 above)
Comment (rey42) [reply src/devices/lacrosse_tx34.c]: done

File: src/devices/lacrosse_tx34.c [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    int events;
+
+    // 22 bits preamble (shifted left): 101010b 0x2DD4
+    static const uint8_t preamble[] = { 0xA8, 0xB7, 0x50 };
+
+    // process rows
+    events = 0;
+    for (row = 0; row <= bitbuffer->num_rows; ++row)
+    {
+        unsigned int start_pos, payload_bits;
```
Comment (zuckschwerdt) [inline src/devices/lacrosse_tx34.c:72]: just `unsigned` and put it directly on bitbuffer_search line

File: src/devices/lacrosse_tx34.c [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+        // search for preamble
+        start_pos = bitbuffer_search(bitbuffer, row, 0, preamble,
+            LACROSSE_TX34_PREAMBLE_BITS);
```
Comment (zuckschwerdt) [inline src/devices/lacrosse_tx34.c:76]: continued lines need indent by 8 spaces

File: src/devices/lacrosse_tx34.c [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        if (start_pos == bitbuffer->bits_per_row[row])
+        continue; // preamble not found
```
Comment (zuckschwerdt) [inline src/devices/lacrosse_tx34.c:78]: indent missing, also below. maybe try clang-format (but don't trust all it's suggestions).

File: src/devices/lacrosse_tx34.c [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+            start_pos + LACROSSE_TX34_PREAMBLE_BITS,
+            payload, LACROSSE_TX34_PAYLOAD_BITS);
+        // verify CRC
+        r_crc = payload[4];
+        c_crc = crc8(&payload[0], 4,
+            LACROSSE_TX34_CRC_POLY, LACROSSE_TX34_CRC_INIT);
+        if (r_crc != c_crc)
+        {
+            // bad CRC: reject IT frame
+            fprintf(stderr,
```
Comment (zuckschwerdt) [inline src/devices/lacrosse_tx34.c:99]: guard with `if (decoder->verbose)`

File: src/devices/lacrosse_tx34.c [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+                "LaCrosse IT frame bad CRC: calculated %02x, "
+                "received %02x\n",
+                c_crc, r_crc);
+            continue;
+        }
+
+        // check model
+        if (((payload[0] & 0xF0) >> 4) != LACROSSE_TX34_ITMODEL)
+            continue; // not a rain gauge...
+        if (decoder->verbose)
```
Comment (zuckschwerdt) [inline src/devices/lacrosse_tx34.c:109]: remove this

**Comment by zuckschwerdt** [comment]
Did you close this by accident? The code looks good.

File: src/devices/lacrosse_tx34.c (hunk 9 above)
Comment (rey42) [reply src/devices/lacrosse_tx34.c:109]: I wont. It would accept frame of every kind of IT+ sensor (thermo/hydro, tea pot...) instead of rain gauges only.

Comment (zuckschwerdt) [reply src/devices/lacrosse_tx34.c:109]: I mean this `if` with `fprintf` below ;)

Comment (rey42) [reply src/devices/lacrosse_tx34.c:108]: Please could you give a reason for this removal? There is same debug print in other decoders (`like lacrosse_tx35.x`).

Comment (zuckschwerdt) [reply src/devices/lacrosse_tx34.c:108]: Well, the next thing we do here is `decoder_output_data()` so `model: LaCrosse TX34-IT` will already be on the screen.
Yes, other decoders need a cleanup for fewer debug prints someday.

Comment (rey42) [reply src/devices/lacrosse_tx34.c:108]: OK

**Comment by rey42** [comment]
Thanks :)
Could add Reynald Poittevin <reynald@poittevin.name> to AUTHORS.

**Comment by zuckschwerdt** [comment]
Thank *you*. The AUTHORS file is historical and not currated anymore. We would like a hall-of-fame for everyone who contributed, but it's rather much work to go through all commits in both repos and history before that and source files, and on top clean it up. Someday ;)
I'll add the mail address to the copyright line.

**Pull request merged**